// bottender/dist/context/MessengerEvent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class MessengerEvent {
    constructor(rawEvent, options = {}) {
        this._rawEvent = rawEvent;
        this._isStandby = options.isStandby || false;
        this._pageId = options.pageId || null;
    }
    get rawEvent() {
        return this._rawEvent;
    }
    get isMessage() {
        return (!!this._rawEvent.message && typeof this._rawEvent.message === 'object');
    }
    get message() {
        return this._rawEvent.message || null;
    }
    get isText() {
        return this.isMessage && typeof this.message.text === 'string';
    }
    get text() {
        if (this.isText) {
            return this.message.text;
        }
        return null;
    }
    get hasAttachment() {
        return (this.isMessage &&
            !!this.message.attachments &&
            this.message.attachments.length > 0);
    }
    get attachments() {
        if (this.message && this.message.attachments) {
            return this.message.attachments;
        }
        return null;
    }
    get isImage() {
        return this.hasAttachment && this.attachments[0].type === 'image';
    }
    get image() {
        return this.isImage ? this.attachments[0].payload : null;
    }
    get isAudio() {
        return this.hasAttachment && this.attachments[0].type === 'audio';
    }
    get audio() {
        return this.isAudio ? this.attachments[0].payload : null;
    }
    get isVideo() {
        return this.hasAttachment && this.attachments[0].type === 'video';
    }
    get video() {
        return this.isVideo ? this.attachments[0].payload : null;
    }
    get isLocation() {
        return (this.hasAttachment && this.attachments[0].type === 'location');
    }
    get location() {
        return this.isLocation ? this.attachments[0].payload : null;
    }
    get isFile() {
        return this.hasAttachment && this.attachments[0].type === 'file';
    }
    get file() {
        return this.isFile ? this.attachments[0].payload : null;
    }
    get isFallback() {
        return (this.hasAttachment && this.attachments[0].type === 'fallback');
    }
    get fallback() {
        return this.isFallback ? this.attachments[0] : null;
    }
    get isSticker() {
        return (this.isMessage && typeof this.message.sticker_id === 'number');
    }
    get sticker() {
        return this.isSticker ? this.message.sticker_id : null;
    }
    get isLikeSticker() {
        return (this.isSticker &&
            (this.message.sticker_id === 369239263222822 ||
                this.message.sticker_id === 369239343222814 ||
                this.message.sticker_id === 369239383222810));
    }
    get isQuickReply() {
        return (this.isMessage &&
            !!this.message.quick_reply &&
            typeof this.message.quick_reply === 'object');
    }
    get quickReply() {
        if (this.message && this.message.quick_reply) {
            return this.message.quick_reply;
        }
        return null;
    }
    get isEcho() {
        return this.isMessage && !!this.message.is_echo;
    }
    get isPostback() {
        return (!!this._rawEvent.postback && typeof this._rawEvent.postback === 'object');
    }
    get postback() {
        return this._rawEvent.postback || null;
    }
    get isGamePlay() {
        return (!!this._rawEvent.game_play && typeof this._rawEvent.game_play === 'object');
    }
    get gamePlay() {
        if (!this.isGamePlay) {
            return null;
        }
        const rawGamePlay = this._rawEvent.game_play;
        let payload;
        try {
            payload = JSON.parse(rawGamePlay.payload);
        }
        catch (e) {
            payload = rawGamePlay.payload;
        }
        return Object.assign(Object.assign({}, rawGamePlay), { payload });
    }
    get isOptin() {
        return !!this._rawEvent.optin && typeof this._rawEvent.optin === 'object';
    }
    get optin() {
        return this._rawEvent.optin || null;
    }
    get isPayment() {
        return (!!this._rawEvent.payment && typeof this._rawEvent.payment === 'object');
    }
    get payment() {
        return this._rawEvent.payment || null;
    }
    get isCheckoutUpdate() {
        return (!!this._rawEvent.checkout_update &&
            typeof this._rawEvent.checkout_update === 'object');
    }
    get checkoutUpdate() {
        return this._rawEvent.checkout_update || null;
    }
    get isPreCheckout() {
        return (!!this._rawEvent.pre_checkout &&
            typeof this._rawEvent.pre_checkout === 'object');
    }
    get preCheckout() {
        return this._rawEvent.pre_checkout || null;
    }
    get isRead() {
        return !!this._rawEvent.read && typeof this._rawEvent.read === 'object';
    }
    get read() {
        return this.isRead ? this._rawEvent.read : null;
    }
    get isDelivery() {
        return (!!this._rawEvent.delivery && typeof this._rawEvent.delivery === 'object');
    }
    get delivery() {
        return this.isDelivery ? this._rawEvent.delivery : null;
    }
    get isPayload() {
        return ((!!this.postback && typeof this.postback.payload === 'string') ||
            (!!this.quickReply && typeof this.quickReply.payload === 'string'));
    }
    get payload() {
        if (!!this.postback && this.isPayload) {
            return this.postback.payload;
        }
        if (!!this.quickReply && this.isPayload) {
            return this.quickReply.payload;
        }
        return null;
    }
    get isPolicyEnforcement() {
        return (!!this._rawEvent['policy-enforcement'] &&
            typeof this._rawEvent['policy-enforcement'] === 'object');
    }
    get policyEnforcement() {
        return this._rawEvent['policy-enforcement'] || null;
    }
    get isAppRoles() {
        return (!!this._rawEvent.app_roles && typeof this._rawEvent.app_roles === 'object');
    }
    get appRoles() {
        return this._rawEvent.app_roles || null;
    }
    get isStandby() {
        return this._isStandby;
    }
    get isPassThreadControl() {
        return (!!this._rawEvent.pass_thread_control &&
            typeof this._rawEvent.pass_thread_control === 'object');
    }
    get passThreadControl() {
        return this._rawEvent.pass_thread_control || null;
    }
    get isTakeThreadControl() {
        return (!!this._rawEvent.take_thread_control &&
            typeof this._rawEvent.take_thread_control === 'object');
    }
    get takeThreadControl() {
        return this._rawEvent.take_thread_control || null;
    }
    get isRequestThreadControl() {
        return (!!this._rawEvent.request_thread_control &&
            typeof this._rawEvent.request_thread_control === 'object');
    }
    get isRequestThreadControlFromPageInbox() {
        return (!!this._rawEvent.request_thread_control &&
            typeof this._rawEvent.request_thread_control === 'object' &&
            this._rawEvent.request_thread_control.requested_owner_app_id ===
                263902037430900);
    }
    get requestThreadControl() {
        return this._rawEvent.request_thread_control || null;
    }
    get isFromCustomerChatPlugin() {
        const isMessageFromCustomerChatPlugin = !!(this.isMessage &&
            !!this.message.tags &&
            this.message.tags.length !== 0 &&
            this.message.tags.some((tag) => tag.source === 'customer_chat_plugin'));
        const isReferralFromCustomerChatPlugin = !!(this.isReferral &&
            this.referral &&
            this.referral.source === 'CUSTOMER_CHAT_PLUGIN');
        return isMessageFromCustomerChatPlugin || isReferralFromCustomerChatPlugin;
    }
    get isReferral() {
        return !!(this._rawEvent.referral ||
            (this._rawEvent.postback && this._rawEvent.postback.referral));
    }
    get referral() {
        if (!this.isReferral) {
            return null;
        }
        return (this._rawEvent.referral ||
            (this._rawEvent.postback && this._rawEvent.postback.referral) ||
            null);
    }
    get ref() {
        if (!this.isReferral) {
            return null;
        }
        return this.referral && this.referral.ref;
    }
    get pageId() {
        return this._pageId || null;
    }
    get isBrandedCamera() {
        return (!!this._rawEvent.branded_camera &&
            typeof this._rawEvent.branded_camera === 'object');
    }
    get brandedCamera() {
        if (!this.isBrandedCamera) {
            return null;
        }
        return this._rawEvent.branded_camera;
    }
}
exports.default = MessengerEvent;
